In [2]:
#library for downloading
!pip install wget

#importing libraries we will use
import numpy as np
import wget
import pandas as pd

#### Question 1
What's the version of NumPy that you installed?

In [3]:
#get version of numpy being used
np.__version__

'1.21.5'

In [4]:


#download data to working directory
!python -m wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv


Saved under data (1).csv


In [5]:
#preview of dataset
df=pd.read_csv("data.csv")
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

#### Question 2
How many records are in the dataset?

In [6]:
df.shape[0]

11914

###### number of rows is 11914

#### Question 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

In [7]:
df['Make'].value_counts()[:3]

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

###### most frequent car manufacturers were Chevrolet, Ford and Volkswagen

#### Question 4
What's the number of unique Audi car models in the dataset?

In [8]:
#get only audi make cars
audis=df.loc[df['Make'] == 'Audi']
audis

Make Model  Year             Engine Fuel Type  Engine HP  \
17     Audi   100  1992             regular unleaded      172.0   
18     Audi   100  1992             regular unleaded      172.0   
19     Audi   100  1992             regular unleaded      172.0   
20     Audi   100  1992             regular unleaded      172.0   
21     Audi   100  1992             regular unleaded      172.0   
...     ...   ...   ...                          ...        ...   
10901  Audi   TTS  2016  premium unleaded (required)      292.0   
10902  Audi   TTS  2017  premium unleaded (required)      292.0   
11178  Audi    V8  1992             regular unleaded      276.0   
11179  Audi    V8  1993             regular unleaded      276.0   
11180  Audi    V8  1994             regular unleaded      276.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
17                  6.0            MANUAL  front wheel drive              4.0   
18                  6.0            MANUAL  front wheel drive              4.0   
19                  6.0         AUTOMATIC    all wheel drive              4.0   
20                  6.0            MANUAL  front wheel drive              4.0   
21                  6.0            MANUAL    all wheel drive              4.0   
...                 ...               ...                ...              ...   
10901               4.0  AUTOMATED_MANUAL    all wheel drive              2.0   
10902               4.0  AUTOMATED_MANUAL    all wheel drive              2.0   
11178               8.0         AUTOMATIC    all wheel drive              4.0   
11179               8.0         AUTOMATIC    all wheel drive              4.0   
11180               8.0         AUTOMATIC    all wheel drive              4.0   

                             Market Category Vehicle Size Vehicle Style  \
17                                    Luxury      Midsize         Sedan   
18                                    Luxury      Midsize         Sedan   
19                                    Luxury      Midsize         Wagon   
20                                    Luxury      Midsize         Sedan   
21                                    Luxury      Midsize         Sedan   
...                                      ...          ...           ...   
10901       Factory Tuner,Luxury,Performance      Compact         Coupe   
10902  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
11178                     Luxury,Performance      Midsize         Sedan   
11179                     Luxury,Performance      Midsize         Sedan   
11180                     Luxury,Performance      Midsize         Sedan   

       highway MPG  city mpg  Popularity   MSRP  
17              24        17        3105   2000  
18              24        17        3105   2000  
19              20        16        3105   2000  
20              24        17        3105   2000  
21              21        16        3105   2000  
...            ...       ...         ...    ...  
10901           27        23        3105  51900  
10902           27        23        3105  52500  
11178           18        13        3105   2000  
11179           18        13        3105   2149  
11180           18        13        3105   2239  

[328 rows x 16 columns]

In [9]:
audis['Model'].unique().shape[0]

34

###### number of unique Audi car models is 34

#### Question 5
How many columns in the dataset have missing values?

In [10]:
null_counts=df.isna().sum()
null_counts

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [11]:
#get Make with null_counts>0
null_above_zero=null_counts.loc[null_counts>0]
null_above_zero

Engine Fuel Type       3
Engine HP             69
Engine Cylinders      30
Number of Doors        6
Market Category     3742
dtype: int64

In [12]:
null_above_zero.shape[0]

5

##### 5 columns in the dataset have missing values

#### Question 6
- Find the median value of "Engine Cylinders" column in the dataset.
- Next, calculate the most frequent value of the same "Engine Cylinders".
- Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
- Now, calculate the median value of "Engine Cylinders" once again.
- Has it changed?

In [13]:
#median value of "Engine Cylinders"
df['Engine Cylinders'].median()

6.0

In [30]:
#the most frequent value of the same "Engine Cylinders"
most_freq_cylinder=df['Engine Cylinders'].mode()
most_freq_cylinder

4752

In [40]:
df["Engine Cylinders"].fillna(most_freq_cylinder[0],inplace=True)


In [41]:
#median value of "Engine Cylinders" after filling up null values
df['Engine Cylinders'].median()

6.0

##### has the median value of "Engine Cylinders": No

#### Question 7

Select all the "Lotus" cars from the dataset.
Select only columns "Engine HP", "Engine Cylinders".
Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
Get the underlying NumPy array. Let's call it X.
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
Invert XTX.
Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
What's the value of the first element of w?

In [55]:
#Select all the "Lotus" cars from the dataset
lotus=df.loc[df['Make']=='Lotus']

#Select only columns "Engine HP", "Engine Cylinders".
select_columns=lotus[['Engine HP','Engine Cylinders']]
select_columns


Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0
3917      189.0               4.0
3918      217.0               4.0
3919      217.0               4.0
4216      350.0               8.0
4217      350.0               8.0
4218      350.0               8.0
4257      400.0               6.0
4258      400.0               6.0
4259      276.0               6.0
4260      276.0               6.0
4261      276.0               6.0
4262      345.0               6.0
4263      345.0               6.0
4264      276.0               6.0
4265      345.0               6.0
4266      276.0               6.0
4267      345.0               6.0
4268      276.0               6.0
4292      257.0               4.0
4293      240.0               4.0
4294      240.0               4.0
4295      240.0               4.0
4296      257.0               4.0
4297      257.0               4.0

In [56]:
select_columns.drop_duplicates()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [61]:
X=select_columns.drop_duplicates()

(9, 2)

In [63]:
XTX=np.dot(X.T,X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [68]:
XTX_inverse=np.linalg.inv(XTX)
y=[1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]
w=np.dot(XTX_inverse,X.T)
w=np.dot(w,y)
w

array([  4.59494481, -63.56432501])

###### the value of the first element of w is 4.5949